include/libdnf/utils/utils.hpp

```diff
@@ -55,6 +58,41 @@ enum class ProblemRules {
     RULE_UNKNOWN
 };
 
+enum class GoalProblem : uint32_t {
+    NO_PROBLEM = 0,
+    SOLVER_ERROR = (1 << 0),
+    REMOVAL_OF_PROTECTED = (1 << 1),
+    NOT_FOUND = (1 << 2),
+    EXCLUDED = (1 << 3),
+    ONLY_SRC = (1 << 4),
+    NOT_FOUND_IN_REPOSITORIES = (1 << 5)
+};
+
+inline GoalProblem operator|(GoalProblem lhs, GoalProblem rhs) {
+    return static_cast<GoalProblem>(
+        static_cast<std::underlying_type<GoalProblem>::type>(lhs) |
+        static_cast<std::underlying_type<GoalProblem>::type>(rhs));
+}
+
+inline GoalProblem operator|=(GoalProblem & lhs, GoalProblem rhs) {
+    lhs = static_cast<GoalProblem>(
+        static_cast<std::underlying_type<GoalProblem>::type>(lhs) |
+        static_cast<std::underlying_type<GoalProblem>::type>(rhs));
+    return lhs;
+}
+
+inline GoalProblem operator&(GoalProblem lhs, GoalProblem rhs) {
+    return static_cast<GoalProblem>(
+        static_cast<std::underlying_type<GoalProblem>::type>(lhs) &
+        static_cast<std::underlying_type<GoalProblem>::type>(rhs));
+}
+
+inline GoalProblem operator-(GoalProblem lhs, GoalProblem rhs) {
+    return static_cast<GoalProblem>(
+        static_cast<std::underlying_type<GoalProblem>::type>(lhs) -
```

Comment:
Interesting implementation of `-`. For example, `ONLY_SRC - NOT_FOUND == EXCLUDED | NOT_FOUND`
You may want to "subtract"/delete flags/bits. No arithmetic subtraction. Maybe this:
```
static_cast<std::underlying_type<GoalProblem>::type>(lhs) &
~static_cast<std::underlying_type<GoalProblem>::type>(rhs));
```

Reply:
I found a similar code:
```
inline QueryCmp operator-(QueryCmp lhs, QueryCmp rhs) {
    return static_cast<QueryCmp>(
        static_cast<std::underlying_type<QueryCmp>::type>(lhs) -
        static_cast<std::underlying_type<QueryCmp>::type>(rhs));
}
```

Reply:
OK. We should fix QueryCmp.